### ISSUE TITLE
EXPECT_ENVOY_BUG behaves inconsistently in release tests

### ISSUE THREAD

**Issue by ravenblackx** [envoyproxy/envoy#29217]
I added three very simple tests in #29176 tagged as death tests and using `EXPECT_ENVOY_BUG`.

Some experimentation reveals that in CI, the third one of these tests fails in bazel.release builds. If the tests are reordered, the third one is still the one that fails (i.e. it is not the specifics of the test case, but the fact that it is third, that provokes failure).

I am now tagging these three tests with GTEST_SKIP and this issue number, so it can be usefully investigated by someone more familiar with the log-based validation behavior of `EXPECT_ENVOY_BUG`.

Coverage is passing without the tests, so skipping them is okay for now. Skipping them only for release builds wouldn't work because they would still fail for coverage builds, and disabling them for release *and* coverage runs would render them mostly pointless since the tests' main value is adding coverage.

**Comment by ravenblackx** [comment]
I have a vague memory of something being changed to cap test log size to avoid overwhelming CI - wonder if these logs might be right on the cusp of that limit, such that a third case gets truncated and therefore can't be found in the log. But again, the whole thing is not something I know much about, so this is just a loose thought, perhaps more likely to be misleading than relevant.

**Comment by alyssawilk** [comment]
yeah I'd say if it's passing locally and failing in CI, it's probably a CI issue.  if you can confirm it fails locally I can try to track down someone to debug

**Comment by alyssawilk** [comment]
oh, also could be the exponential backoff on envoy_bug - I think @abeyad helped Matt debug something like this recently?

**Comment by alyssawilk** [comment]
oops @adisuissa not @abeyad (misread firstname)

**Comment by adisuissa** [comment]
Thanks Alyssa, Indeed seems to be relevant.

ENVOY_BUG in release mode (`-c opt`) uses exponential triggering, implemented by [this code](https://github.com/envoyproxy/envoy/blob/1b0803556cba74657b5d0ce7fef8454b255996dd/source/common/common/assert.cc#L88).
The [call](https://github.com/envoyproxy/envoy/blob/1b0803556cba74657b5d0ce7fef8454b255996dd/source/common/common/assert.h#L272) just passes the ENVOY_BUG line number.

**Comment by ravenblackx** [comment]
Yep, that seems like almost certainly it then, since the ENVOY_BUG in question is the same one for different paths that lead to it. Not logging on the third case would therefore make perfect sense, since 1 and 2 are both powers of two.

I think a good fix (in that it would prevent this from surprising other people, versus manually clearing would work in one context but leave the same trap for next time) might be to include `EnvoyBugState::get().clear()` in the behavior of `EXPECT_ENVOY_BUG`, what do you think?
